File: app/src/main/java/com/umbrella/budgetapp/Customs/Spinners.java
package com.umbrella.budgetapp.Customs;

import android.content.res.TypedArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.umbrella.budgetapp.Database.Collections.Account;
import com.umbrella.budgetapp.Database.Collections.Currency;
import com.umbrella.budgetapp.Extensions.App;
import com.umbrella.budgetapp.R;

import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public class Spinners {

    /**
     * Private binder for the Spinner widget.
     * @param spinner The spinner that is used for the collection.
     * @param adapter The adapter with the collection of items.
     */
    private static void bindSpinner(Spinner spinner, ArrayAdapter<?> adapter) {
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);
    }

    /**
     * Public DocumentReference of the selected item. Mostly used when saving changes in a Fragment.
     * @param spinner The spinner that has to be checked.
     * @return The DocumentReference of the selected item.
     */
    public static DocumentReference getReference(Spinner spinner) {
        return ((QueryDocumentSnapshot)spinner.getSelectedItem()).getReference();
    }

    /**
     * Class for the color spinner.
     */
    public static class Colors {

        public enum Size {
            SMALL,
            LARGE
        }

        public Colors(@NonNull Fragment fragment, @NonNull Spinner spinner, @Nullable Size size) {
            List<Integer> currentList = new LinkedList<>();
            for (int i: fragment.getResources().getIntArray(R.array.colors)) currentList.add(i);

            bindSpinner(spinner, new ColorAdapter(fragment, currentList, (size != null ? size : Size.SMALL)));
        }

        private static class ColorAdapter extends ArrayAdapter<Integer> {
            private Fragment fragment;
            private List<Integer> items;
            private Size size;

            ColorAdapter(Fragment fragment, List<Integer> objects, Size size) {
                super(fragment.requireContext(), R.layout.custom_spinner_colors, 0, objects);
                this.fragment = fragment;
                this.items = objects;
                this.size = size;
            }

            @Override
            public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            @NonNull
            @Override
            public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            private View createItemView(int position, ViewGroup parent) {
                final View view = LayoutInflater.from(fragment.getContext()).inflate(R.layout.custom_spinner_colors, parent, false);

                View item = view.requireViewById(R.id.spinner_color);
                item.setLayoutParams(new LinearLayout.LayoutParams((size == Size.LARGE ? fragment.getResources().getDimensionPixelSize(R.dimen.spinner_colors_width_large) : fragment.getResources().getDimensionPixelSize(R.dimen.spinner_colors_width_small)), fragment.getResources().getDimensionPixelSize(R.dimen.spinners_colors_height)));
                item.setForeground(fragment.getResources().getDrawable(R.drawable._spinner_color_corners, getDropDownViewTheme()));
                item.setBackgroundColor(items.get(position));
                return view;
            }
        }
    }

    /**
     * Class for the icon spinner.
     */
    public static class Icons {
        private  static Integer[] size;

        public Icons(@NonNull Fragment fragment, @NonNull Spinner spinner) {
           size = new Integer[fragment.getResources().getIntArray(R.array.icons).length];
           bindSpinner(spinner, new IconAdapter(fragment));
        }

        private static class IconAdapter extends ArrayAdapter<Integer> {
            private Fragment fragment;
            private TypedArray array;

            IconAdapter(Fragment fragment) {
                super(fragment.requireContext(), R.layout.custom_spinner_icons, size);
                this.fragment = fragment;
                this.array = fragment.getResources().obtainTypedArray(R.array.icons);
            }

            @Override
            public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            @NonNull
            @Override
            public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            private View createItemView(int position, ViewGroup parent) {
                final View view = LayoutInflater.from(fragment.getContext()).inflate(R.layout.custom_spinner_icons, parent, false);

                ((ImageView)view.requireViewById(R.id.spinner_icon)).setImageResource(array.getResourceId(position,-1));
                return view;
            }
        }
    }

    /**
     * Class for the account spinner.
     */
    public static class Accounts {
        public static List<QueryDocumentSnapshot> list = new LinkedList<>();

        public Accounts(@NonNull Fragment fragment, @NonNull Spinner spinner) {
            App.getInstance().getUserDocument().collection(Account.COLLECTION).orderBy(Account.POSITION, Query.Direction.ASCENDING).get().addOnSuccessListener(hasAccounts -> {
                List<QueryDocumentSnapshot> snapList = new LinkedList<>();
                hasAccounts.forEach(snapList::add);

                list = snapList;
                bindSpinner(spinner, new AccountsAdapter(fragment, snapList));
            });
        }

        public int getPosition(DocumentReference id) {
            return list.stream().filter(item -> item.getReference().equals(id))
                    .collect(Collectors.toList()).get(0).toObject(Currency.class)
                    .getPosition();
        }

        public static void initializeAndSetDefault(Fragment fragment, Spinner spinner, DocumentReference id) {
            App.getInstance().getUserDocument().collection(Account.COLLECTION).orderBy(Account.POSITION, Query.Direction.ASCENDING).get().addOnSuccessListener(hasAccounts -> {
                List<QueryDocumentSnapshot> snapList = new LinkedList<>();
                hasAccounts.forEach(snapList::add);

                list = snapList;
                bindSpinner(spinner, new AccountsAdapter(fragment, snapList));

                spinner.setSelection(list.stream().filter(item -> item.getReference().equals(id))
                        .collect(Collectors.toList()).get(0).toObject(Currency.class)
                        .getPosition());
            });
        }

        private static class AccountsAdapter extends ArrayAdapter<QueryDocumentSnapshot> {
            private Fragment fragment;
            private List<QueryDocumentSnapshot> snaps;

            AccountsAdapter(Fragment fragment, List<QueryDocumentSnapshot> accounts) {
                super(fragment.requireContext(), R.layout.custom_spinner_items, accounts);
                this.fragment = fragment;
                this.snaps = accounts;
            }

            @Override
            public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            @NonNull
            @Override
            public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            private View createItemView(int position, ViewGroup parent) {
                final View view = LayoutInflater.from(fragment.getContext()).inflate(R.layout.custom_spinner_items, parent, false);

                ((TextView)view.requireViewById(R.id.custom_spinner_item)).setText(snaps.get(position).toObject(Account.class).getName());
                return view;
            }
        }
    }

    /**
     * Class for the account spinner.
     */
    public static class Currencies {
        public static List<QueryDocumentSnapshot> list = new LinkedList<>();

        public Currencies(@NonNull Fragment fragment, @NonNull Spinner spinner) {
            App.getInstance().getUserDocument().collection(Currency.COLLECTION).orderBy(Currency.POSITION, Query.Direction.ASCENDING).get().addOnSuccessListener(hasCurrencies -> {
                List<QueryDocumentSnapshot> snapList = new LinkedList<>();
                hasCurrencies.forEach(snapList::add);

                list = snapList;
                bindSpinner(spinner, new CurrenciesAdapter(fragment, snapList));
            });
        }

        public int getPosition(DocumentReference id) {
            return list.stream().filter(item -> item.getReference().equals(id))
                    .collect(Collectors.toList()).get(0).toObject(Currency.class)
                    .getPosition();
        }

        public static void initializeAndSetDefault(Fragment fragment, Spinner spinner, DocumentReference id) {
            App.getInstance().getUserDocument().collection(Currency.COLLECTION).orderBy(Currency.POSITION, Query.Direction.ASCENDING).get().addOnSuccessListener(hasCurrencies -> {
                List<QueryDocumentSnapshot> snapList = new LinkedList<>();
                hasCurrencies.forEach(snapList::add);

                list = snapList;
                bindSpinner(spinner, new CurrenciesAdapter(fragment, snapList));

                spinner.setSelection(list.stream().filter(item -> item.getReference().equals(id))
                        .collect(Collectors.toList()).get(0).toObject(Currency.class)
                        .getPosition());
            });
        }

        private static class CurrenciesAdapter extends ArrayAdapter<QueryDocumentSnapshot> {
            private Fragment fragment;
            private List<QueryDocumentSnapshot> snaps;

            CurrenciesAdapter(Fragment fragment, List<QueryDocumentSnapshot> currencies) {
                super(fragment.requireContext(), R.layout.custom_spinner_items, currencies);
                this.fragment = fragment;
                this.snaps = currencies;
            }

            @Override
            public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            @NonNull
            @Override
            public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                return createItemView(position, parent);
            }

            private View createItemView(int position, ViewGroup parent) {
                final View view = LayoutInflater.from(fragment.getContext()).inflate(R.layout.custom_spinner_items, parent, false);

                ((TextView)view.requireViewById(R.id.custom_spinner_item)).setText(snaps.get(position).toObject(Currency.class).getCountryID());
                return view;
            }

        }
    }
}

File: app/src/main/java/com/umbrella/budgetapp/ui/customs/Spinners.kt
package com.umbrella.budgetapp.ui.customs

import android.content.res.TypedArray
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import com.umbrella.budgetapp.R
import com.umbrella.budgetapp.database.collections.Account
import com.umbrella.budgetapp.database.collections.subcollections.CurrencyAndName
import com.umbrella.budgetapp.database.defaults.DefaultCountries
import com.umbrella.budgetapp.database.viewmodels.AccountViewModel
import com.umbrella.budgetapp.database.viewmodels.CurrencyViewModel
import com.umbrella.budgetapp.extensions.inflate

sealed class Spinners {
    /**
     * Private binder for the Spinner widget.
     * @param spinner The spinner that is used for the collection.
     * @param adapter The adapter with the collection of items.
     */
    internal fun bindSpinner(spinner: Spinner, adapter: ArrayAdapter<*>) {
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter
    }

    /**
     * Class for the color spinner.
     */
    class Colors(fragment: Fragment, spinner: Spinner, size: Size) : Spinners() {
        enum class Size {
            SMALL,
            LARGE
        }

        init {
            bindSpinner(spinner, ColorAdapter(fragment, fragment.resources.getIntArray(R.array.colors).asList(), size))
        }

        private class ColorAdapter(private val fragment: Fragment, private val items: List<Int>, private val size: Size) : ArrayAdapter<Int>(fragment.requireContext(), R.layout.custom_spinner_colors, 0, items) {

            override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
                return createItemView(position, parent)
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                return createItemView(position, parent)
            }

            private fun createItemView(position: Int, parent: ViewGroup): View {
                val view = parent.inflate(R.layout.custom_spinner_colors, false)
                val item = view.findViewById<View>(R.id.spinner_color)
                item.layoutParams = LinearLayout.LayoutParams(if (size == Size.LARGE) fragment.resources.getDimensionPixelSize(R.dimen.spinner_colors_width_large) else fragment.resources.getDimensionPixelSize(R.dimen.spinner_colors_width_small), fragment.resources.getDimensionPixelSize(R.dimen.spinners_colors_height))
                item.foreground = ResourcesCompat.getDrawable(fragment.resources, R.drawable._spinner_color_corners, dropDownViewTheme)
                item.setBackgroundColor(items[position])
                return view
            }

        }
    }

    /**
     * Class for the icon spinner.
     */
    class Icons(fragment: Fragment, spinner: Spinner) : Spinners() {

        init {
            bindSpinner(spinner, IconAdapter(fragment, fragment.resources.obtainTypedArray(R.array.icons), (0..fragment.resources.getIntArray(R.array.icons).count()).toList().dropLast(1)))
        }

        private class IconAdapter(private val fragment: Fragment, private val array: TypedArray, indices: List<Int>) : ArrayAdapter<Int>(fragment.requireContext(), R.layout.custom_spinner_icons, 0, indices) {

            override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
                return createItemView(position, parent)
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                return createItemView(position, parent)
            }

            private fun createItemView(position: Int, parent: ViewGroup): View {
                val view = parent.inflate(R.layout.custom_spinner_icons, false)
                (view.findViewById<View>(R.id.spinner_icon) as ImageView).setImageResource(array.getResourceId(position, 0))
                return view
            }


        }
    }

    /**
     * Class for simple spinners of Strings.
     */
    class SimpleSpinner(fragment: Fragment, spinner: Spinner, arrayId: Int) : Spinners() {

        init {
            val items : Array<String> = fragment.resources.getStringArray(arrayId)

            bindSpinner(spinner, ArrayAdapter<String>(fragment.requireContext(), android.R.layout.simple_spinner_item, items))
        }
    }

    /**
     * Internal function for initializing the adapter and binding it to the adapter.
     */
    internal fun <T> initRoom(fragment: Fragment, spinner: Spinner) : ArrayAdapter<T> {
        val adapter = ArrayAdapter<T>(fragment.requireContext(), android.R.layout.simple_spinner_item, mutableListOf())

        bindSpinner(spinner, adapter)

        return adapter
    }

    /**
     * Class for the currency spinner.
     */
    class Currencies(fragment: Fragment, spinner: Spinner) : Spinners() {

        private var list = mutableListOf<CurrencyAndName>()

        init {
            val model by fragment.viewModels<CurrencyViewModel>()

            val adapter = initRoom<String>(fragment, spinner)

            model.getBasicCurrencies().observe(fragment.viewLifecycleOwner, Observer { items ->
                list = items.toMutableList()

                val defC = DefaultCountries()
                val listOfNames = mutableListOf<String>()
                items.forEach { item -> listOfNames.add(defC.getCountryById(item.countryRef).name) }

                adapter.apply {
                    clear()
                    addAll(listOfNames)
                    notifyDataSetChanged()
                }
            })

            spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    spinner.tag = list[position]
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }

        fun getPosition(currencyId: Long) : Int {
            return list.indexOf(list.find { curr -> curr.id == currencyId })
        }
    }

    /**
     * Class for the account spinner.
     */
    class Accounts(fragment: Fragment, spinner: Spinner) : Spinners() {

        private var list = mutableListOf<Account>()

        init {
            val model by fragment.viewModels<AccountViewModel>()

            val adapter = initRoom<String>(fragment, spinner)

            model.getAccountBasics().observe(fragment.viewLifecycleOwner, Observer { items ->
                list = items.toMutableList()

                val listOfNames = mutableListOf<String>()
                items.forEach { item -> listOfNames.add(item.name!!) }

                adapter.apply {
                    clear()
                    addAll(listOfNames)
                    notifyDataSetChanged()
                }

                spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                        spinner.tag = list[position]
                    }

                    override fun onNothingSelected(parent: AdapterView<*>?) {}
                }
            })
        }

        fun getPosition(accountId: Long) : Int {
            return list.indexOf(list.find { acc -> acc.id == accountId })
        }
    }
}